src/index.ts

```diff
@@ -1582,86 +1494,112 @@ async function createClaimXChainPayload(
-	}
-
-	config.verbose && console.log('Squid route calculated :)')
-	// config.verbose && console.log('Full route : ', route)
+	config.verbose && console.log('Squid route calculated :)', { route })
 
 	// cryptography
-
-	const paramsHash = ethers.utils.solidityKeccak256(
-		['address', 'address', 'bytes', 'uint256'],
-		[recipient, transactionRequest.target, transactionRequest.data, transactionRequest.value]
-	) // this works
-
-	async function _signMessage(message: ethers.utils.Bytes, privateKey: string) {
-		const wallet = new ethers.Wallet(privateKey)
-		const signature = await wallet.signMessage(message)
-		console.log(`${signature} (Created by _signMessage without hash)`)
-		return signature
-	}
-
-	const signature = await _signMessage(ethers.utils.arrayify(paramsHash), keys.privateKey)
-
-	// log all returns
-	const result = {
-		recipientAddress: recipient,
-		tokenAmount: tokenAmount,
-		hash: paramsHash,
-		signature: signature,
-		idx: depositIdx,
-		chainId: params.chainId,
-		contractVersion: params.contractVersion,
-		params: params,
-		estimate: estimate,
-		transactionRequest: transactionRequest,
+	const routerContractVersion = 'R' + linkDetails.contractVersion
+	const squidAddress = isMainnet ? SQUID_ADDRESS['mainnet'] : SQUID_ADDRESS['testnet']
+	const vaultAddress = getContractAddress(linkDetails.chainId, linkDetails.contractVersion)
+	const routerAddress = getContractAddress(linkDetails.chainId, routerContractVersion)
+	const normalWithdrawalPayload = await createClaimPayload(link, routerAddress, true)
+
+	const routingArgs = [
+		'0x1900',
+		routerAddress,
+		linkDetails.chainId,
+		vaultAddress,
+		linkDetails.depositIndex,
+		squidAddress,
+		route.value,
+		0, // currently we are not charging any fees
+		route.calldata,
+	]
+	console.log('Routing args', routingArgs)
+
+	const packedRoutingData = ethers.utils.solidityPack(
+		[
+			'bytes2', // 0x1900 as per EIP-191
+			'address', // peanut router address
+			'uint256', // chain id
+			'address', // peanut vault address
+			'uint256', // deposit index
+			'address', // squid address
+			'uint256', // squid fee
+			'uint256', // peanut fee
+			'bytes', // squid calldata
+		],
+		routingArgs
+	)
+	console.log('Packed routing data %s', packedRoutingData)
+
+	const xchainDigest = ethers.utils.solidityKeccak256(['bytes'], [packedRoutingData])
+	console.log('X chain digest', xchainDigest)
+
+	// const signingKey = new SigningKey(keys.privateKey)
+	const wallet = new ethers.Wallet(keys.privateKey)
+	const routingSignatureRaw = await wallet._signingKey().signDigest(ethers.utils.arrayify(xchainDigest))
+	const routingSignature = routingSignatureRaw.r + routingSignatureRaw.s.slice(2) + routingSignatureRaw.v.toString(16)
+	config.verbose && console.log(`Peanut routing signature:`, { routingSignature: routingSignatureRaw })
+
+	const result: interfaces.IXchainClaimPayload = {
+		chainId,
+		contractVersion: routerContractVersion,
```

Comment:
just a note here, the 'contractVersion' term was initially thought for the peanut contract. v1, v2, v2.1, etc. Sequential

Now we're introducing the router contract and batcher contracts. That muddles the term contractVersion a little bit. Not sure if there is an elegant alternative though.

Reply:
Yeah, that's true. I think that as long as we can deterministically parse the contract version, it should be fine.

One thing though that is not really convenient right now to do is to compare contract versions. Say I want to specify that a certain function works only with 4.2+ contracts. Rn there is no easy way to do it. Maybe we could try using an enum for contract versions, but not sure whether it will really help.

Reply:
agreed. let's keep an eye on this and consider dealing with it if we add more versions.